### PR TITLE
Prototype theme foundations

### PR DESCRIPTION
**Pull request by kawaijoe**
This PR lays the foundation of how `FTheme` should work.

### CODE REVIEW

File: forui/lib/src/box/box.dart [hunk 1]
```diff
@@ -0,0 +1,28 @@
+import 'package:flutter/material.dart';
+
+import 'package:forui/forui.dart';
+
+/// A sample widget demonstrating the architecture of ForUI.
```
Comment (Pante) [inline forui/lib/src/box/box.dart:5]:
```suggestion
/// A sample widget that demonstrates Forui's architecture.
```

File: forui/lib/src/box/box_style.dart [hunk 2]
```diff
@@ -0,0 +1,19 @@
+import 'package:flutter/material.dart';
+
+import 'package:forui/forui.dart';
+
+/// A class that holds the styles for [FBox].
```
Comment (Pante) [inline forui/lib/src/box/box_style.dart:5]:
```suggestion
/// [FBox]'s style.
```

File: forui/lib/src/box/box_style.dart [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+  /// The color.
+  final Color color;
+
+  /// The text.
+  final TextStyle text;
+
+  /// Creates a [FBoxStyle].
+  const FBoxStyle({required this.color, required this.text});
+
+  /// Creates a [FBoxStyle] that inherits the properties from the given [FFontData].
```
Comment (Pante) [inline forui/lib/src/box/box_style.dart:16]:
```suggestion
  /// Creates a [FBoxStyle] that inherits its properties from [data] and [style].
```

File: forui/lib/src/theme/font/font_data.dart [hunk 4]
```diff
@@ -0,0 +1,32 @@
+import 'package:flutter/material.dart';
+
+/// A class that holds the font data.
```
Comment (Pante) [inline forui/lib/src/theme/font/font_data.dart:3]:
```suggestion
/// The default font data that is inherited from a [FTheme] by child Forui widgets.
```

I think we should further explain what the fontSize & other fields are used for and how they are used to scale text in widgets.

File: forui/lib/src/theme/font/font_data.dart [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+  final double fontSize;
+
+  /// The letter spacing that corresponds to the `letterSpacing` property on [TextStyle].
+  final double letterSpacing;
+
+  /// The word spacing that corresponds to the `wordSpacing` property on [TextStyle].
+  final double wordSpacing;
+
+  /// The height that corresponds to the `height` property on [TextStyle].
+  final double height;
```
Comment (Pante) [inline forui/lib/src/theme/font/font_data.dart:15]: I feel like these names are really misleading. Shouldn't they be scaling factors?

Also  we should first explain that these properties are used for multiplicative scaling before then mentioning that it is similar to a property in TextStyle in a 2nd sentence.

File: forui/lib/src/theme/font/text_style_builder.dart [hunk 6]
```diff
@@ -0,0 +1,17 @@
+import 'package:flutter/material.dart';
+
+import 'package:forui/forui.dart';
+
+/// An extension that adds a `inherit` constructor to [TextStyleBuilder].
+extension type TextStyleBuilder(TextStyle style) implements TextStyle {
```
Comment (Pante) [inline forui/lib/src/theme/font/text_style_builder.dart:6]: Thinking about it, maybe `ScaledTextStyle` would be a better name.

File: forui/lib/src/theme/font/text_style_builder.dart [hunk 7]
```diff
@@ -0,0 +1,17 @@
+import 'package:flutter/material.dart';
+
+import 'package:forui/forui.dart';
+
+/// An extension that adds a `inherit` constructor to [TextStyleBuilder].
```
Comment (Pante) [inline forui/lib/src/theme/font/text_style_builder.dart:5]:
```suggestion
/// A [TextStyle] that scales its properties by factors.
/// 
/// The scaled properties are:
/// * Font size
/// * Height
/// * Letter spacing
/// * Word spacing
```

File: forui/lib/src/theme/font/text_style_builder.dart (hunk 6 above)
Comment (Pante) [inline forui/lib/src/theme/font/text_style_builder.dart:6]:
```suggestion
extension type TextStyleBuilder._(TextStyle style) implements TextStyle {
```

File: forui/lib/src/theme/font/text_style_builder.dart [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+extension type TextStyleBuilder(TextStyle style) implements TextStyle {
+  /// Creates a [TextStyle] that inherits the properties from the given [FFontData].
+  TextStyleBuilder.inherit(FFontData data, TextStyle style)
+      : style = TextStyle(
+    fontSize: _calculateFactor(style.fontSize, data.fontSize),
+    letterSpacing: _calculateFactor(style.letterSpacing, data.letterSpacing),
+    wordSpacing: _calculateFactor(style.wordSpacing, data.wordSpacing),
+    height: _calculateFactor(style.height, data.height),
+  );
```
Comment (Pante) [inline forui/lib/src/theme/font/text_style_builder.dart:14]:
```suggestion
  TextStyleBuilder(TextStyle base, FFontData data): style = TextStyle(
    fontSize: _calculateFactor(base.fontSize, data.fontSize),
    letterSpacing: _calculateFactor(base.letterSpacing, data.letterSpacing),
    wordSpacing: _calculateFactor(base.wordSpacing, data.wordSpacing),
    height: _calculateFactor(base.height, data.height),
  );
```

File: forui/lib/src/theme/style/zinc_style.dart [hunk 9]
```diff
@@ -7,34 +7,40 @@ import 'package:forui/forui.dart';
 /// The zinc style is based on [shadcn/ui](https://ui.shadcn.com/themes).
 extension FZincThemeData on Never {
```
Comment (Pante) [inline forui/lib/src/theme/style/zinc_style.dart:8]: nitpick: I think the file name should be changed to `zinc_theme.dart`.

File: forui/lib/src/theme/style/zinc_style.dart [hunk 10]
```diff
@@ -7,34 +7,40 @@ import 'package:forui/forui.dart';
 /// The zinc style is based on [shadcn/ui](https://ui.shadcn.com/themes).
```
Comment (Pante) [inline forui/lib/src/theme/style/zinc_style.dart:7]:
```suggestion
/// The zinc theme, based on [shadcn/ui](https://ui.shadcn.com/themes).
```

File: forui/lib/src/theme/style/zinc_style.dart [hunk 11]
```diff
@@ -7,34 +7,40 @@ import 'package:forui/forui.dart';
...
 extension FZincThemeData on Never {
   /// The light theme of [FZincThemeData].
```
Comment (Pante) [inline forui/lib/src/theme/style/zinc_style.dart:9]:
```suggestion
  /// The light variant.
```

File: forui/lib/src/theme/style/zinc_style.dart [hunk 12]
```diff
@@ -7,34 +7,40 @@ import 'package:forui/forui.dart';
...
+      muted: const Color(0xFFF4F4F5),
+      mutedForeground: const Color(0xFF71717A),
+      destructive: const Color(0xFFEF4444),
+      destructiveForeground: const Color(0xFFFAFAFA),
+      border: const Color(0xFFE4E4E7),
+      borderRadius: BorderRadius.circular(8),
+    ),
   );
 
   /// The dark theme of [FZincThemeData].
```
Comment (Pante) [inline forui/lib/src/theme/style/zinc_style.dart:28]:
```suggestion
  /// The dark variant.
```

File: forui/lib/src/theme/style_data.dart [hunk 13]
```diff
@@ -0,0 +1,86 @@
+import 'package:flutter/material.dart';
+
+/// A class that holds the theme data for the app.
```
Comment (Pante) [inline forui/lib/src/theme/style_data.dart:3]:
```suggestion
/// The default theme data that is inherited from a [FTheme] by child Forui widgets.
```

File: forui/lib/src/theme/theme_data.dart [hunk 14]
```diff
@@ -1,86 +1,31 @@
 import 'package:flutter/material.dart';
 
+import 'package:forui/forui.dart';
+
 /// A class that holds the theme data for the app.
```
Comment (Pante) [inline forui/lib/src/theme/theme_data.dart:5]:
```suggestion
/// The default font and theme data that are inherited by child Forui widgets.
```

File: forui/lib/src/theme/widget_data.dart [hunk 15]
```diff
@@ -0,0 +1,21 @@
+import 'package:forui/forui.dart';
+
+/// A class that holds the widget data.
```
Comment (Pante) [inline forui/lib/src/theme/widget_data.dart:3]:
```suggestion
/// The default widget-specific theme and font data that is inherited from a [FTheme] by child Forui widgets.
```

**Review by Pante — CHANGES_REQUESTED** [review]
I think the layout of the widgets is good, the documentation just needs to be revised slightly.

File: forui/lib/src/theme/font/font_data.dart (hunk 5 above)
Comment (kawaijoe) [reply forui/lib/src/theme/font/font_data.dart:18]: That's a good catch. I decided to suffix the variables with `Scalar` as it is similar to definition in physics and scalar vector multiplication in linear algebra.

File: forui/lib/src/theme/style/zinc_style.dart (hunk 9 above)
Comment (Pante) [reply forui/lib/src/theme/style/zinc_style.dart:8]: Maybe `forui/lib/src/themes/zinc_theme,dart`?

File: forui/lib/src/theme/font/font_data.dart [hunk 16]
```diff
@@ -0,0 +1,43 @@
+import 'package:flutter/material.dart';
+
+import 'package:forui/forui.dart';
+
+/// The font data that is inherited from a [FTheme] deigned to be utilized by [ScaledTextStyle] and child Forui widgets.
```
Comment (Pante) [inline forui/lib/src/theme/font/font_data.dart:5]:
```suggestion
/// The font data that is inherited from a [FTheme] by child Forui widgets and [ScaledTextStyle]s.
```

File: forui/lib/src/theme/font/font_data.dart [hunk 17]
```diff
@@ -0,0 +1,43 @@
...
+///
+/// This class contains scalar values that are multiplied with the corresponding properties on [TextStyle] to ensure
+/// that various fonts are scaled consistently across the application.
+class FFontData {
+  /// The font size scalar that multiplies with the `fontSize` property on [TextStyle].
```
Comment (Pante) [inline forui/lib/src/theme/font/font_data.dart:10]:
```suggestion
  /// A value used to scale [TextStyle.fontSize].
```

File: forui/lib/src/theme/font/font_data.dart [hunk 18]
```diff
@@ -0,0 +1,43 @@
...
+  final double fontSizeScalar;
+
+  /// The letter spacing scalar that multiplies with the `letterSpacing` property on [TextStyle].
```
Comment (Pante) [inline forui/lib/src/theme/font/font_data.dart:13]:
```suggestion
  /// /// A value used to scale [TextStyle.letterSpacing].
```

**Review by Pante — APPROVED** [review]

**Pull request merged**